pt/blockchain/helper.py

```diff
@@ -0,0 +1,245 @@
+                price_buy = buy_bids[i][1]
+                break
+        else:
+            price_buy = buy_bids[-1][1]
+
+        # volume sell
+        for i in range(1, len(sell_bids)):
+            if sell_bids[i - 1][0] < volume <= sell_bids[i][0]:
+                price_sell = sell_bids[i - 1][1]
+                break
+        else:
+            price_sell = sell_bids[-1][1]
+
+        if price_buy <= price_sell:
+            matched = True
+            break
+
+        matched_price = price_sell
+        matched_volume = volume
+
+    if matched:
+        return matched_volume, matched_price
+    return None
+
+
+def generate_success_matchresult(start_time, tx_hash, win_value, win_price):
+    buy_bidsubmits, sell_bidsubmits = get_matched_bidsubmits(start_time, win_price)
+    buy_total = sum([bid.value for bid in buy_bidsubmits])
+    sell_total = sum([bid.value for bid in sell_bidsubmits])
+
+    while buy_bidsubmits and sell_bidsubmits:
+        buy_volume = round(win_value * (buy_bidsubmits[-1].value / buy_total), 2)
+        sell_volume = round(win_value * (sell_bidsubmits[-1].value / sell_total), 2)
+        if buy_volume > sell_volume:
+            # update both bidsubmit for sell and buy
+            update_bidsubmit(buy_bidsubmits[-1], sell_bidsubmits[-1], tx_hash, sell_volume, win_price)
+
+            # clone all columns except uuid
+            data = buy_bidsubmits[-1].__dict__
+            if data.pop("uuid", None):
+                data.pop("_sa_instance_state", None)
+                buy_bidsubmits[-1] = BidSubmit.add(**data)
+
+            # sell allocated, pop the bid from behind
+            sell_bidsubmits.pop()
+        elif buy_volume == sell_volume:
+            # update both bidsubmit for sell and buy
+            update_bidsubmit(buy_bidsubmits[-1], sell_bidsubmits[-1], tx_hash, sell_volume, win_price)
+
+            # sell and buy both allocated, pop them from behind
+            buy_bidsubmits.pop()
+            sell_bidsubmits.pop()
+        else:
+            # update both bidsubmit for sell and buy
+            update_bidsubmit(buy_bidsubmits[-1], sell_bidsubmits[-1], tx_hash, buy_volume, win_price)
+
+            # clone all columns except uuid
+            data = sell_bidsubmits[-1].__dict__
+            if data.pop("uuid", None):
+                data.pop("_sa_instance_state", None)
+                sell_bidsubmits[-1] = BidSubmit.add(**data)
+            buy_bidsubmits.pop()
+
+
+def update_bidsubmit(buy, sell, tx_hash, value, price):
+    buyer = buy.tenders.user
+    seller = sell.tenders.user
+
+    # buy_bidsubmit
+    buy.win = 1
+    buy.status = "已得標"
+    buy.counterpart_name = seller.username
+    buy.counterpart_address = seller.address
+    buy.win_value = value
+    buy.win_price = price
+    buy.achievement = 0
+    buy.settlement = 0
+    buy.transaction_hash = tx_hash
+    buy.upload_time = datetime.now()
+    buy.update()
+
+    # buy_bidsubmit
```

Comment:
**sell**_bidsubmit

Reply:
fixed in 0fbdf07d9504f6aec2d2e3228e2d016c4cf81363

Reply:
fixed in 0fbdf07d9504f6aec2d2e3228e2d016c4cf81363